### PR TITLE
[MONO] Move marshal-ilgen into a component

### PR DESCRIPTION
**Pull request by naricc**
This PR moves marshal-ilgen into a component. This will save approximately 100 KB in the runtime where the component is not needed. Further refactoring should allow more code to be moved into the component in the future.

This also makes a change to the CMake files so that some components can be statically linked in the aot compiler, since marshal_ilgen is needed at AOT time.

Contributes to: https://github.com/dotnet/runtime/issues/61685

### CODE REVIEW

**Comment by srxqds** [comment]
so when should we use this component or not?

**Comment by naricc** [comment]
@srxqds Basically whenever you have `DisableRuntimeMarshallingAttribute` set in a project, and publish that project, we would not have to link the component into the published.  A very quick summary is that the `DisableRuntimeMarshallingAttribute` says not to use the old runtime marshaling, but instead use source generator marshaling. 

More information here: https://github.com/dotnet/runtime/issues/60639 and here: https://github.com/dotnet/runtime/issues/61685

**Comment by MichalStrehovsky** [comment]
> Basically whenever you have `DisableRuntimeMarshallingAttribute` set in a project, and publish that project, we would not have to link the component into the published

Nit: one would also have to make sure all of the referenced NuGets have that attribute (or no pinvoke, delegate, or unmanaged Calli), Assembly.LoadFrom and similar is not called, and nobody Reflection.Emits a pinvoke.

**Comment by naricc** [comment]
Some of the test failures are this: https://github.com/dotnet/runtime/issues/72016

File: src/mono/mono/component/CMakeLists.txt [hunk 1]
```diff
@@ -78,17 +81,51 @@ set(${MONO_DIAGNOSTICS_TRACING_COMPONENT_NAME}-dependencies
...
+# marshal_ilgen
+list(APPEND components
+  ${MONO_MARSHAL_ILGEN_COMPONENT_NAME}
+)
+list(APPEND components_for_aot
+  ${MONO_MARSHAL_ILGEN_COMPONENT_NAME}
+)
+
+set(${MONO_MARSHAL_ILGEN_COMPONENT_NAME}-sources
+  ${MONO_COMPONENT_PATH}/marshal_ilgen.c
```
Comment (vargaz) [inline src/mono/mono/component/CMakeLists.txt:93]: Mono traditionally uses - as a separator in file names not _.

File: src/mono/mono/component/marshal_ilgen.h [hunk 2]
```diff
@@ -4,6 +4,17 @@
 
 #include "metadata/marshal-lightweight.h"
 #include "metadata/marshal.h"
+#include "mono/component/component.h"
+
+typedef struct MonoComponentMarshalIlgen {
+	MonoComponent component;
```
Comment (vargaz) [inline src/mono/mono/component/marshal_ilgen.h:10]: The 'L' should be uppercased in MonoComponentMarshalIlgen.

File: src/mono/mono/metadata/marshal.c [hunk 3]
```diff
@@ -127,6 +129,66 @@ static GENERATE_TRY_GET_CLASS_WITH_CACHE (unmanaged_callconv_attribute, "System.
...
+	&mono_marshal_shared_get_sh_dangerous_add_ref,
+	&mono_marshal_shared_get_sh_dangerous_release,
+	&mono_marshal_shared_init_safe_handle,
+	&mono_marshal_shared_is_in,
+	&mono_marshal_shared_is_out,
+	&mono_marshal_shared_mb_emit_exception_marshal_directive,
+	&mono_mb_add_local,
+	&mono_mb_emit_add_to_local,
+	&mono_mb_emit_auto_layout_exception,
+	&mono_mb_emit_branch,
```
Comment (vargaz) [inline src/mono/mono/metadata/marshal.c:162]: Why are these callbacks, can't the component call these directly ?

Comment (naricc) [reply src/mono/mono/metadata/marshal.c:162]: Components can't call directly to mono, except for functions marked MONO_API or MONO_COMPONENT_API. I think direct calls might actually work under some linking scenarios, but not all of them.

File: src/mono/mono/component/CMakeLists.txt (hunk 1 above)
Comment (naricc) [reply src/mono/mono/component/CMakeLists.txt:93]: I was under the impression that the component name "marshal_ilgen" also gets expanded into C variable, so I can't put a dash in it. Maybe I am mistaken and I can change the name of the file independently; I will try that.

Comment (naricc) [reply src/mono/mono/component/CMakeLists.txt:93]: It looks like it can be made to build with a - instead of a _, but all the other component names (and files with component names in them) use _. This also makes it consistent with the _ for component names  the C code. Should I use _ to be consistent with the other components? Or change it to be -?

Comment (lambdageek) [reply src/mono/mono/component/CMakeLists.txt:93]: let's use `-`. we can cleanup the other components file names (but probably not the component names - they're used by the SDKs) in a separate PR. (I think it's probably my fault - I named `hot_reload` like that because it matched the symbol name)

Comment (naricc) [reply src/mono/mono/component/CMakeLists.txt]: Ok, done.

**Comment by naricc** [comment]
@vargaz @lambdageek I think this is ready to go in, unless you have any additional feedback.

File: src/mono/mono/component/marshal-ilgen-stub.c [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+static bool 
+marshal_ilgen_available (void)
+{
+	return false;
+}
+
+static int
+stub_emit_marshal_ilgen (EmitMarshalContext *m, int argnum, MonoType *t,
+	      MonoMarshalSpec *spec, int conv_arg,	
+	      MonoType **conv_arg_type, MarshalAction action, MonoMarshalLightweightCallbacks* lightweigth_cb)
```
Comment (radical) [inline src/mono/mono/component/marshal-ilgen-stub.c:15]: nit: mix of tabs, spaces

File: src/mono/mono/component/marshal-ilgen-stub.c [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+{
+	return 0;
+}
+
+static void 
+mono_component_marshal_ilgen_stub_init(void)
+{
+}
```
Comment (radical) [inline src/mono/mono/component/marshal-ilgen-stub.c:23]:
```suggestion
}
```

File: src/mono/mono/component/marshal-ilgen-stub.c [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+static MonoComponentMarshalILgen component_func_table = {
+	{ MONO_COMPONENT_ITF_VERSION, &marshal_ilgen_available },
+	mono_component_marshal_ilgen_stub_init,
+	stub_emit_marshal_ilgen,
+	stub_mono_marshal_ilgen_install_callbacks_mono
+};
+
+MonoComponentMarshalILgen*
+ mono_component_marshal_ilgen_init (void)
```
Comment (radical) [inline src/mono/mono/component/marshal-ilgen-stub.c:38]:
```suggestion
mono_component_marshal_ilgen_init (void) 
```

File: src/mono/mono/component/marshal-ilgen.c [hunk 7]
```diff
@@ -0,0 +1,2859 @@
...
+#include "metadata/method-builder-ilgen.h"
+#include "metadata/custom-attrs-internals.h"
+#include "metadata/class-init.h"
+#include "mono/metadata/class-internals.h"
+#include "metadata/reflection-internals.h"
+#include "mono/metadata/handle.h"
+#include "mono/component/component.h"
+
+
+
```
Comment (radical) [inline src/mono/mono/component/marshal-ilgen.c:15]: nit: unnecessary empty lines

File: src/mono/mono/component/marshal-ilgen.c [hunk 8]
```diff
@@ -0,0 +1,2859 @@
...
+		}
+	}
+
+	return TRUE;
+}
+
+static int
+emit_marshal_ptr_ilgen (EmitMarshalContext *m, int argnum, MonoType *t,
+		  MonoMarshalSpec *spec, int conv_arg,
+		  MonoType **conv_arg_type, MarshalAction action)
```
Comment (radical) [inline src/mono/mono/component/marshal-ilgen.c:869]: spaces -> tabs
```suggestion
			MonoMarshalSpec *spec, int conv_arg,
			MonoType **conv_arg_type, MarshalAction action)
```

File: src/mono/mono/component/marshal-ilgen.c [hunk 9]
```diff
@@ -0,0 +1,2859 @@
...
+		break;
+	}
+	return conv_arg;
+}
+
+static int
+emit_marshal_boolean_ilgen (EmitMarshalContext *m, int argnum, MonoType *t,
+		      MonoMarshalSpec *spec,
+		      int conv_arg, MonoType **conv_arg_type,
+		      MarshalAction action)
```
Comment (radical) [inline src/mono/mono/component/marshal-ilgen.c:902]: tab/spaces mix

Comment (radical) [reply src/mono/mono/component/marshal-ilgen.c:902]: There are are more. I would suggest fixing that for all the changes here.

Comment (naricc) [reply src/mono/mono/component/marshal-ilgen.c]: I didn't actually make marshal-ilgen in this change, its just been moved. But I guess it makes sense to fix these while I am at it.

File: src/mono/wasm/build/WasmApp.Native.targets [hunk 10]
```diff
@@ -29,7 +29,10 @@
   </ItemGroup>
 
   <ItemGroup Condition="'$(Configuration)' == 'Debug' and '@(_MonoComponent->Count())' == 0">
-    <_MonoComponent Include="hot_reload;debugger" />
+    <_MonoComponent Include="hot_reload;debugger;marshal-ilgen" />
+  </ItemGroup>
+  <ItemGroup Condition="'$(Configuration)' == 'Release' and '@(_MonoComponent->Count())' == 0">
+    <_MonoComponent Include="marshal-ilgen" />
```
Comment (radical) [inline src/mono/wasm/build/WasmApp.Native.targets:35]: Just curious, why is this needed by default for Release?

Comment (naricc) [reply src/mono/wasm/build/WasmApp.Native.targets:35]: Right now, we always need the component (in both Debug and Release). The tooling to drop it when not needed has not been implemented yet, its part three of this issue: https://github.com/dotnet/runtime/issues/61685. 

I guess I could add a branch to here to *always* add it, regardless of Release/Debug, if that would make it clearer.

Comment (radical) [reply src/mono/wasm/build/WasmApp.Native.targets:35]: > I guess I could add a branch to here to always add it, regardless of Release/Debug, if that would make it clearer.

I think that would be good.

**Comment by radical** [comment]
I don't really know much about the actual changes, so I can't review anything else 😬

File: src/mono/wasm/build/WasmApp.Native.targets (hunk 10 above)
Comment (naricc) [reply src/mono/wasm/build/WasmApp.Native.targets:35]: Looking at it again, I am not sure if it is that easy. What is this `@(_MonoComponent->Count())' == 0` condition that was originally there is for? Presumably we only want to add items to _MonoComponent *once* for some reason, but why?

Comment (radical) [reply src/mono/wasm/build/WasmApp.Native.targets:35]: I think we are setting those as the default, so if nobody else has added `_MonoComponent`, then we use these defaults. @lambdageek can answer it best

File: src/mono/mono/metadata/marshal-noilgen.c [hunk 11]
```diff
@@ -3,7 +3,8 @@
 #include <mono/metadata/attrdefs.h>
 #include "metadata/marshal-internals.h"
 #include "metadata/marshal.h"
-#include "metadata/marshal-ilgen.h"
+// TODO: PRobably shouldn't directly refer to a header in a component like this?
+#include "component/marshal-ilgen.h"
 #include "utils/mono-compiler.h"
```
Comment (lambdageek) [inline src/mono/mono/metadata/marshal-noilgen.c:8]: 2 things:

1. all of these should be `mono/metadata/xyz.h` not `metadata/xyz.h`
2. I really don't like that `mono/component/marshal-ilgen.h` contains: (a) declarations of functions that aren't in the component; (b) declarations of _noilgen_ functions.  Please make two separate headers.

**Review by lambdageek — APPROVED** [review]
LGTM, but please split up `component/marshal-ilgen.h` into two headers so the non-component stuff isn't declared in teh component header

**Comment by lambdageek** [comment]
@naricc when this is merged, please open issues for xamarin-macios and xamarin-android to make sure they include the marshal-ilgen component by default.

/cc @jonathanpeppers @rolfbjarne

**Comment by jonathanpeppers** [comment]
We'll also need to know what MSBuild property toggles the inclusion of this component. We have these basically hardcoded right now:

https://github.com/xamarin/xamarin-android/blob/73f10df63b00ee8a21d61183056a26922c83e89d/src/Xamarin.Android.Build.Tasks/Microsoft.Android.Sdk/targets/Microsoft.Android.Sdk.AssemblyResolution.targets#L198-L200

**Comment by lambdageek** [comment]
> We'll also need to know what MSBuild property toggles the inclusion of this component. We have these basically hardcoded right now:
> 
> https://github.com/xamarin/xamarin-android/blob/73f10df63b00ee8a21d61183056a26922c83e89d/src/Xamarin.Android.Build.Tasks/Microsoft.Android.Sdk/targets/Microsoft.Android.Sdk.AssemblyResolution.targets#L198-L200

@jonathanpeppers @naricc There are going to be two phases to this:

1. The mobile SDKs should just hardcode always using the `marshal-ilgen` component.  For net7 this is the conservative backward-compatible behavior and will basically maintain the same functionality we have today.
2. Separately we will need to do some work to decide when to _not_ include `marshal-ilgen`.  This will bring some space savings to apps that fully convert to `LibraryImportGenerator` (and to nugets that all use LibraryImportGenerator, and to reflection emit that only generates blittable pinvokes, etc).  Probably as a first step we will try this out in the runtime and in Wasm and once we have some experience we will work on a design together with the mobile SDKs (and probably CoreCLR-based scenarios like single file, or native AOT).  I doubt this is going to be in 7.

tl;dr hardcoding the component to always be used is exactly what we want to do for 7.

**Comment by jonathanpeppers** [comment]
Ok, great. We will probably put something like:
```xml
<_MonoComponent Condition=" '$(_IncludeMarshalIlGen)' != 'false' " Include="marshal-ilgen" />
```
Then you could experiment with `_IncludeMarshalIlGen=false`. but it's always there by default.

**Comment by radical** [comment]
Merge `main` to get fixes for some of the failing wasm jobs.

**Comment by naricc** [comment]
Catalyst failure might be this, although it is a different test: https://github.com/dotnet/runtime/issues/72840

**Comment by naricc** [comment]
wasm issue appears to be this: https://github.com/dotnet/runtime/issues/73419

**Comment by radical** [comment]
- How is this getting tested?
- Should `System.Diagnostics.Tracing` tests be run for AOT? It is disabled right now.
- How does a user trigger this? Is there a msbuild property? Also, if that is set then will it trigger relinking when building the project (and not publishing)?

**Comment by naricc** [comment]
> * How is this getting tested?

The marshaling code itself is not really getting changed, just moved into a component. And we are running it throuhg all the test suites to make sure that component gets included in every scenario.

> * Should `System.Diagnostics.Tracing` tests be run for AOT? It is disabled right now.

I don't know. @mdh1418 Should it?

> * How does a user trigger this? Is there a msbuild property? Also, if that is set then will it trigger relinking when building the project (and not publishing)?

This PR doesn't yet have any way to trigger it; the component is always included for now. The next step of this work is to have a way to drop the component when it is not needed. The plan is to have some property set in the project, along with an anlyzer that checks that `DisableRuntimeMarshaling` is set on all the referenced projects (as well as the project itself).

File: eng/testing/tests.mobile.targets [hunk 12]
```diff
@@ -74,6 +74,119 @@
...
+    <Message Importance="High" Text="PackageId:       $(ApkPackageId)"/>
+    <Message Importance="High" Text="Instrumentation: net.dot.MonoRunner"/>
+    <Message Importance="High" Text="Apk:             $(ApkBundlePath)"/>
+
+    <Copy SourceFiles="$(ApkBundlePath)"
+          DestinationFolder="$(TestArchiveTestsDir)"
+          SkipUnchangedFiles="true"
+          Condition="'$(ArchiveTests)' == 'true' and '$(IgnoreForCI)' != 'true'" />
+  </Target>
+
```
Comment (mdh1418) [inline eng/testing/tests.mobile.targets:189]: I think we moved the android parts into `tests.android.targets` `AndroidApp.*.targets` `AndroidApp.*.props` in https://github.com/dotnet/runtime/pull/72394, so any changes for marshal-ilgen into the Android targets should be in those instead.

File: src/mono/msbuild/android/build/AndroidApp.targets [hunk 13]
```diff
@@ -119,8 +119,14 @@
     </ItemGroup>
   </Target>
 
+
```
Comment (mdh1418) [inline src/mono/msbuild/android/build/AndroidApp.targets:122]: remove extra lines?

File: src/mono/msbuild/android/build/AndroidApp.targets [hunk 14]
```diff
@@ -119,8 +119,14 @@
...
+
   <Target Name="_AndroidGenerateAppBundle" DependsOnTargets="_AndroidGenerateRuntimeConfig">
 
+    <PropertyGroup>
+      <RuntimeComponents Condition="'$(RuntimeComponents)' == ''" >marshal-ilgen</RuntimeComponents>
```
Comment (mdh1418) [inline src/mono/msbuild/android/build/AndroidApp.targets:127]: Would we not want to add `marshal-ilgen` to `RuntimeComponents` if it contained something else? When do we want to include `marshal-ilgen`?

File: src/mono/msbuild/apple/build/AppleApp.targets [hunk 15]
```diff
@@ -109,9 +109,16 @@
     </ItemGroup>
   </Target>
 
+
```
Comment (mdh1418) [inline src/mono/msbuild/apple/build/AppleApp.targets:112]: remove extra lines?

File: src/mono/msbuild/apple/build/AppleApp.targets [hunk 16]
```diff
@@ -109,9 +109,16 @@
...
   </Target>
 
+  
+
   <Target Name="_AppleGenerateAppBundle" DependsOnTargets="_AppleGenerateRuntimeConfig">
     <!-- Run App bundler, it uses AOT libs (if needed), link all native bits, compile simple UI (written in ObjC)
          and produce an app bundle (with xcode project) -->
+
+    <PropertyGroup>
+      <RuntimeComponents Condition="'$(RuntimeComponents)' == ''">marshal-ilgen</RuntimeComponents>
```
Comment (mdh1418) [inline src/mono/msbuild/apple/build/AppleApp.targets:119]: Same question w.r.t. non empty RuntimeComponents

File: src/mono/sample/Android/AndroidSampleApp.csproj [hunk 17]
```diff
@@ -78,6 +78,7 @@
         <Output TaskParameter="CompiledAssemblies" ItemName="BundleAssemblies" />
     </MonoAOTCompiler>
 
+
```
Comment (mdh1418) [inline src/mono/sample/Android/AndroidSampleApp.csproj:81]: remove extra line?

File: src/mono/sample/iOS/Program.csproj [hunk 18]
```diff
@@ -67,6 +67,10 @@
         <Output TaskParameter="CompiledAssemblies" ItemName="BundleAssemblies" />
     </MonoAOTCompiler>
 
+    <PropertyGroup>
+      <RuntimeComponents Condition="'$(RuntimeComponents)' == ''">marshal-ilgen</RuntimeComponents>
```
Comment (mdh1418) [inline src/mono/sample/iOS/Program.csproj:71]: Same question w.r.t. non empty RuntimeComponents

File: src/mono/wasm/build/WasmApp.Native.targets [hunk 19]
```diff
@@ -28,9 +28,13 @@
     <UpToDateCheckInput Include="@(NativeFileReference)" />
   </ItemGroup>
 
+
```
Comment (mdh1418) [inline src/mono/wasm/build/WasmApp.Native.targets:31]: remove extra line?

File: src/tests/build.proj [hunk 20]
```diff
@@ -230,6 +230,10 @@
         SourceFiles="@(TestTargetingPathLibs)"
         DestinationFolder="$(BuildDir)" />
 
+    <PropertyGroup>
+      <RuntimeComponents Condition="'$(RuntimeComponents)' == ''" >marshal-ilgen</RuntimeComponents>
+    </PropertyGroup>
+
```
Comment (mdh1418) [inline src/tests/build.proj:236]: This seems superfluous if we have L188

**Comment by mdh1418** [comment]
> > * Should `System.Diagnostics.Tracing` tests be run for AOT? It is disabled right now.
> 
> I don't know. @mdh1418 Should it?
> 

Ah, I thought they were re-enabled in https://github.com/dotnet/runtime/pull/72545 from one of the earlier commits but looks like I missed that a later commit only re-enabled it for iOS (should have reenabled for tvOS too). That doesn't seem to change anything testing wise on CI since we only run System.Runtime.Tests.csproj for iOS due to many timeouts in the past from overloading iOS capacity, and we generally do all of the testing on tvOS.

I'm not sure why it is disabled for Browser wasm.

File: src/mono/msbuild/android/build/AndroidApp.targets (hunk 14 above)
Comment (naricc) [reply src/mono/msbuild/android/build/AndroidApp.targets:127]: Because if RuntimeComponents is already set, we don't want to overwrite it, and delete whatever else some one wanted to include. I tried making RuntimeComponents into an ItemGroup instead, so we could just append it. But that did not work for msbuild reasons I haven't been able to understand yet.

Comment (mdh1418) [reply src/mono/msbuild/android/build/AndroidApp.targets:127]: We can avoid overwriting what was originally there via
`<RuntimeComponents>$(RuntimeComponents);marshal-ilgen</RuntimeComponents>`

Comment (lambdageek) [reply src/mono/msbuild/android/build/AndroidApp.targets:127]: I don't think we should use concatenation.
If someone is setting RuntimeComponents explicitly, they should have to specify all of them - including marshal-ilgen if they want it.  (For example testing that we can leave it out will want to leave it out...)

Remember this `RuntimeComponents` property is something we invented in dotnet/runtime and it's not shipping.  Users (ie the mobile and wasm SDKs) set `_MonoComponent` and then look at `_MonoRuntimeComponentLink` or `_MonoRuntimeComponentDontLink` depending on whether they are doing opt-in or opt-out.  So it's only for in-tree testing (using our mobile AppBuilders, or the in-tree wasmApp) that uses this thing.  We can keep it simple to allow testing of weird scenarios.

**Comment by mdh1418** [comment]
https://github.com/dotnet/runtime/pull/73581 will check if the suite passes on atleast tvOS

**Comment by naricc** [comment]
Looks like the wasm errors are this: https://github.com/dotnet/runtime/issues/73666

**Comment by radical** [comment]
`main` has fixes for some of the failing jobs.

**Comment by naricc** [comment]
The wasm debugger tests are known issues, so I am going to merge this.

**Comment by lambdageek** [comment]
@naricc please open issues for XA and XI to add the component by default!

**Comment by naricc** [comment]
@lambdageek 
https://github.com/xamarin/xamarin-android/issues/7249
https://github.com/xamarin/xamarin-macios/issues/15668

**Comment by uweigand** [comment]
When running the test suite on s390x, I just saw the following assertion:
```
Assertion at /home/uweigand/runtime/src/mono/mono/component/marshal-ilgen.c:63, condition `!ilgen_cb_inited' not met
```
with this call stack:
```
  #14 0x000003ffb327d0ac in mono_assertion_message (file=0x3ffb33865a0 "/home/uweigand/runtime/src/mono/mono/component/marshal-ilgen.c", line=63, condition=0x3ffb33865e0 "!ilgen_cb_inited") at /home/uweigand/runtime/src/mono/mono/eglib/goutput.c:226
  #15 0x000003ffb3321a30 in mono_install_marshal_callbacks_ilgen (cb=0x3ff9a3f16d0) at /home/uweigand/runtime/src/mono/mono/component/marshal-ilgen.c:63
  #16 0x000003ffb33221ce in mono_marshal_ilgen_init () at /home/uweigand/runtime/src/mono/mono/component/marshal-ilgen.c:2858
  #17 0x000003ffb3322106 in get_marshal_cb () at /home/uweigand/runtime/src/mono/mono/component/marshal-ilgen.c:2753
  #18 0x000003ffb3321f24 in mono_emit_marshal_ilgen (m=0x3ff9a3f1bc8, argnum=0, t=0x3ff8ad98b68, spec=0x0, conv_arg=0, conv_arg_type=0x3ff8ad99ed8, action=MARSHAL_ACTION_CONV_IN, lightweigth_cb=0x3ffb33910e8 <marshal_lightweight_cb>) at /home/uweigand/runtime/src/mono/mono/component/marshal-ilgen.c:2810
  #19 0x000003ffb2e54c52 in mono_emit_marshal (m=0x3ff9a3f1bc8, argnum=0, t=0x3ff8ad98b68, spec=0x0, conv_arg=0, conv_arg_type=0x3ff8ad99ed8, action=MARSHAL_ACTION_CONV_IN) at /home/uweigand/runtime/src/mono/mono/metadata/marshal.c:3226
  #20 0x000003ffb2f219aa in emit_native_wrapper_ilgen (image=0x2aa18a8b950, mb=0x3ff7c90eb50, sig=0x3ff8ad99cc8, piinfo=0x3ff8ad98af0, mspecs=0x3ff7c8f47e0, func=0x3ff82d81750 <CompressionNative_DeflateInit2_>, flags=(EMIT_NATIVE_WRAPPER_CHECK_EXCEPTIONS | EMIT_NATIVE_WRAPPER_RUNTIME_MARSHALLING_ENABLED)) at /home/uweigand/runtime/src/mono/mono/metadata/marshal-lightweight.c:937
  #21 0x000003ffb2e5755a in mono_marshal_emit_native_wrapper (image=0x2aa18a8b950, mb=0x3ff7c90eb50, sig=0x3ff8ad99cc8, piinfo=0x3ff8ad98af0, mspecs=0x3ff7c8f47e0, func=0x3ff82d81750 <CompressionNative_DeflateInit2_>, flags=(EMIT_NATIVE_WRAPPER_CHECK_EXCEPTIONS | EMIT_NATIVE_WRAPPER_RUNTIME_MARSHALLING_ENABLED)) at /home/uweigand/runtime/src/mono/mono/metadata/marshal.c:6300
  #22 0x000003ffb2e56b6c in mono_marshal_get_native_wrapper (method=0x3ff8ad98af0, check_exceptions=1, aot=0) at /home/uweigand/runtime/src/mono/mono/metadata/marshal.c:3672
  #23 0x000003ffb30328e8 in compile_special (method=0x3ff8ad98af0, error=0x3ff9a3f2c38) at /home/uweigand/runtime/src/mono/mono/mini/mini-runtime.c:2469
  #24 0x000003ffb30322b0 in mono_jit_compile_method_with_opt (method=0x3ff8ad98af0, opt=374417919, jit_only=0, error=0x3ff9a3f2c38) at /home/uweigand/runtime/src/mono/mono/mini/mini-runtime.c:2672
  #25 0x000003ffb3031c04 in jit_compile_method_with_opt_cb (arg=0x3ff9a3f28e8) at /home/uweigand/runtime/src/mono/mono/mini/mini-runtime.c:2762
  #26 0x000003ffb3029ea8 in jit_compile_method_with_opt (params=0x3ff9a3f28e8) at /home/uweigand/runtime/src/mono/mono/mini/mini-runtime.c:2778
  #27 0x000003ffb3029c7e in mono_jit_compile_method (method=0x3ff8ad98af0, error=0x3ff9a3f2c38) at /home/uweigand/runtime/src/mono/mono/mini/mini-runtime.c:2797
  #28 0x000003ffb31677ba in common_call_trampoline (regs=0x3ff9a3f2db8, code=0x3ffa806234c "\a\340\300\001", m=0x3ff8ad98af0, vt=0x0, vtable_slot=0x0, error=0x3ff9a3f2c38) at /home/uweigand/runtime/src/mono/mono/mini/mini-trampolines.c:618
  #29 0x000003ffb3166b8e in mono_magic_trampoline (regs=0x3ff9a3f2db8, code=0x3ffa806234c "\a\340\300\001", arg=0x3ff8ad98af0, tramp=0x3ffb3de02c8 "\300X") at /home/uweigand/runtime/src/mono/mono/mini/mini-trampolines.c:759
```

This seems to be a race condition that is not easily reproducible.  But looking at the code:
```
static MonoMarshalILgenCallbacks *
get_marshal_cb (void)
{
        if (G_UNLIKELY (!ilgen_cb_inited)) {
#ifdef ENABLE_ILGEN
                mono_marshal_ilgen_init ();
#else
                mono_marshal_noilgen_init_heavyweight ();
#endif
        }
        return &ilgen_marshal_cb;
}
```
and 
```
void
mono_install_marshal_callbacks_ilgen (MonoMarshalILgenCallbacks *cb)
{
        g_assert (!ilgen_cb_inited);
        g_assert (cb->version == MONO_MARSHAL_CALLBACKS_VERSION);
        memcpy (&ilgen_marshal_cb, cb, sizeof (MonoMarshalILgenCallbacks));
        ilgen_cb_inited = TRUE;
}
```
there does indeed appear to be a race condition with accessing the `ilgen_cb_inited` flag without any locks.

I'm not sure if this patch is directly responsible - there seems to have been a race condition before, but maybe with the component rework it is now somehow easier to hit? - but I never saw this assertion fail before.
